### PR TITLE
Add id to MetadataKey so that we can change the key order

### PR DESCRIPTION
**Pull request by Jackie-Jiang**
Assign an id to each `MetadataKey` in the `DataTable` so that we can change the key order (insert new keys) if necessary for readability. There have been multiple issues caused by people adding keys in the middle

### CODE REVIEW

**Comment by siddharthteotia** [comment]
Part of me thinks that people really need to read the code (javadoc just above the enum) before making bad changes of inserting keys in the middle. So, we should try to stick to current mechanism if possible.

File: pinot-common/src/main/java/org/apache/pinot/common/utils/DataTable.java [hunk 1]
```diff
@@ -90,64 +90,72 @@ enum MetadataValueType {
...
 
-    // getByOrdinal returns an enum key for a given ordinal or null if the key does not exist.
+    /**
+     * Returns the MetadataKey for the given id, or {@code null} if the id does not exist.
+     */
     @Nullable
-    public static MetadataKey getByOrdinal(int ordinal) {
-      return VALUES[Math.min(ordinal, VALUES.length - 1)];
+    public static MetadataKey getById(int id) {
+      return id < ID_TO_ENUM_KEY_MAP.length ? ID_TO_ENUM_KEY_MAP[id] : null;
```
Comment (richardstartin) [inline pinot-common/src/main/java/org/apache/pinot/common/utils/DataTable.java:147]: This appears to be an unnecessary change as there is a null element at the end of the map to deal with unmapped ordinals.

**Comment by richardstartin** [comment]
> Part of me thinks that people really need to read the code (javadoc just above the enum) before making bad changes of inserting keys in the middle. So, we should try to stick to current mechanism if possible.

+1. Enums should be treated as append-only code constructs.

**Comment by FelixGV** [comment]
> > Part of me thinks that people really need to read the code (javadoc just above the enum) before making bad changes of inserting keys in the middle. So, we should try to stick to current mechanism if possible.
> 
> +1. Enums should be treated as append-only code constructs.

Following conventions is a good convention, but making the code structure enforce or incentivize conventions may be even better.

IMHO, it's not too controversial to add a manually maintained ID for enums that need to be sent across processes or persisted across restarts of the same process (i.e. any enum that will potentially be produced by one version of the software and consumed by another).

Using the enum's built-in ordinal is error-prone, since it doesn’t survive re-ordering, nor insertions and removals from the middle. Using a manually-maintained ID on the other hand is resilient to all these scenarios.

File: pinot-common/src/main/java/org/apache/pinot/common/utils/DataTable.java (hunk 1 above)
Comment (Jackie-Jiang) [reply pinot-common/src/main/java/org/apache/pinot/common/utils/DataTable.java]: Functionality wise, it doesn't make a difference (I updated it to handle negative input as well). Readability wise, IMO it is more readable without extra overhead. It does take me some time to understand that we append a null element to the end of the array.

**Comment by Jackie-Jiang** [comment]
@siddharthteotia @richardstartin
To add some context here, we had got multiple PRs merged without following the convention, which broke the backward compatibility. It is hard for all the contributors and reviewers to remember the convention, especially when the change is not next to the comment describing how to keep backward compatible. As @FelixGV pointed out, using enum's ordinal is error-prone, and we need to use a way to enforce people to follow the convention, and make it easy for reviewers to catch when the convention is not followed.

Example PRs that breaks the backward compatibility:
#9092 Merged then reverted the next day
#8884 Merged but not caught the backward compatibility issue until this PR (already one month!). I actually pointed it out during the [review](https://github.com/apache/pinot/pull/8884#discussion_r905529608), but the fix goes to the wrong class and I didn't get a chance to review it again. This is a proof that we don't have any guard for this convention, except for the comment that is 30 lines away from the actual change.

**Comment by richardstartin** [comment]
> @siddharthteotia @richardstartin To add some context here, we had got multiple PRs merged without following the convention, which broke the backward compatibility. It is hard for all the contributors and reviewers to remember the convention, especially when the change is not next to the comment describing how to keep backward compatible. As @FelixGV pointed out, using enum's ordinal is error-prone, and we need to use a way to enforce people to follow the convention, and make it easy for reviewers to catch when the convention is not followed.
> 
> Example PRs that breaks the backward compatibility: #9092 Merged then reverted the next day #8884 Merged but not caught the backward compatibility issue until this PR (already one month!). I actually pointed it out during the [review](https://github.com/apache/pinot/pull/8884#discussion_r905529608), but the fix goes to the wrong class and I didn't get a chance to review it again. This is a proof that we don't have any guard for this convention, except for the comment that is 30 lines away from the actual change.

Clearly a very simple regression test can prevent changes like this from getting merged (and from someone accidentally adding a duplicate id, against which there are currently no guards). Code review shouldn’t focus on things which can be automated simply because hardly anyone has (or should have) the attention span to verify trivialities like these.

File: pinot-common/src/main/java/org/apache/pinot/common/utils/DataTable.java (hunk 1 above)
Comment (richardstartin) [reply pinot-common/src/main/java/org/apache/pinot/common/utils/DataTable.java]: Ok, it’s better now it supports negative values

**Comment by Jackie-Jiang** [comment]
> Clearly a very simple regression test can prevent changes like this from getting merged (and from someone accidentally adding a duplicate id, against which there are currently no guards). Code review shouldn’t focus on things which can be automated simply because hardly anyone has (or should have) the attention span to verify trivialities like these.

We may add a regression test to verify all the existing keys still have the same ordinal, but we also need to take extra effort maintaining the test whenever a new key is added. E.g. #8884 can still happen if the previous PR didn't add the new keys to the test

We do check duplicate id in the static block.

Assigning an id has extra benefits though:
- People can change the order of the keys for readability (and that's the main reason why people tend to add new keys in the middle so that similar metadata are grouped together, e.g. `numConsumingSegmentsQueried`, `numConsumingSegmentsProcessed`, `numConsumingSegmentsMatched`)
- We may remove the old keys that no longer apply

**Comment by siddharthteotia** [comment]
Let me go through the change once today/tomorrow

**Review by siddharthteotia — APPROVED** [review]
I am ok with this to make things more resilient against some of the problems caused by few recent PRs.

However, if as a reviewer we come across a change (e.g reordering enums entries (without a stronger reason)), then I suggest we still push back / advocate for adding entries at the end and other hygiene / practices.
